File: cpp/43_左旋转字符串.cpp
#include <iostream>
#include <string>

class Solution {
public:
    // 利用stl求解
    std::string LeftRotateString(std::string str, int n) {
        if (str.empty() || n == 0) {
            return str;
        }
        return str.substr(n, str.size() - n) + str.substr(0, n);
    }

    // 高效解法：两部分分别翻转，然后整体翻转
    void str_reverse(std::string& str, int l, int h) {
        while (l < h) {
            char tmp = str[l];
            str[l] = str[h];
            str[h] = tmp;
            l++;
            h--;
        }
    }

    std::string LeftRotateString2(std::string str, int n) {
        if (str.empty() || n == 0) {
            return str;
        }

        str_reverse(str, 0, n - 1);
        str_reverse(str, n, str.size() - 1);
        str_reverse(str, 0, str.size() - 1);
        return str;
    }

};

int main(int argc, char* argv[])
{
    Solution s;
    std::string str = "abcXYZdef";

    std::cout << str << std::endl;
    std::cout << s.LeftRotateString(str, 3) << std::endl;
    std::cout << s.LeftRotateString2(str, 3) << std::endl;
    return 0;
}

File: python/43_左旋转字符串.py
#!/bin/bash python3
#-*- coding:utf-8 -*-

class Solution():
    def LeftRotateString(self, s, n):
        """
        借助下标操作的解法
        """
        if s == "" or n == 0:
            return s
        
        return s[n:] + s[0:n]

    def str_reverse(self, s, l ,h):
        """
        对s[l:h]实现翻转
        """
        tmp = list(s)
        while l < h:
            tmp[l], tmp[h] = tmp[h], tmp[l]
            l += 1
            h -= 1
        s = str(tmp)

    def LeftRotateString2(self, s, n):
        if s == "" or n == 0:
            return s
        
        # 没有对输入的s进行保护
        self.str_reverse(s, 0, n - 1)
        self.str_reverse(s, n, len(s) - 1)
        self.str_reverse(s, 0, len(s) - 1)
        return s


def main():
    s = Solution()
    ss = "absXYZdef"

    print(ss)
    print(s.LeftRotateString(ss, 3))
    print(s.LeftRotateString2(ss, 3))


if __name__ == "__main__":
    main()